File: go/aln/parser.go
package aln

import (
	"bytes"
	"fmt"
)

// Parser buffers a byte sequence and emits packets as they are read
// from a an input stream.

// Packet parsing state enumeration
const (
	STATE_BUFFERING = 0
	STATE_ESCAPED   = 1
)

type PacketCallback func(*Packet) bool
type OnCloseCallback func(Channel)

type Parser struct {
	buffer         *bytes.Buffer
	state          byte // enumerated State
	packetCallback PacketCallback
}

func NewParser(cb PacketCallback) *Parser {
	return &Parser{
		state:          STATE_BUFFERING,
		buffer:         bytes.NewBuffer([]byte{}),
		packetCallback: cb,
	}
}

// Clear resets the state of the parser following a frame
func (p *Parser) Clear() {
	p.buffer.Reset()
	p.state = STATE_BUFFERING
}

func (p *Parser) acceptPacket() bool {
	defer p.Clear()
	if pkt, err := ParsePacket(p.buffer.Bytes()); err == nil {
		return p.packetCallback(pkt)
	} else {
		fmt.Println("on acceptPacket, ParsePacket:" + err.Error())
	}
	return true
}

func (p *Parser) IngestStream(buffer []byte) bool {
	for _, msg := range buffer {
		if msg == FRAME_END {
			if !p.acceptPacket() {
				return false
			}
		} else if p.state == STATE_ESCAPED {
			switch msg {
			case FRAME_END_T:
				p.buffer.WriteByte(FRAME_END)
			case FRAME_ESC_T:
				p.buffer.WriteByte(FRAME_ESC)
			default:
				fmt.Printf("unexpected char in KISS frame: %x", msg)
			}
			p.state = STATE_BUFFERING
		} else if msg == FRAME_ESC {
			p.state = STATE_ESCAPED
		} else {
			p.buffer.WriteByte(msg)
		}
	}
	return true
}

File: python/src/alnmeshpy/parser.py
# Parser class can compose and decompose packets for frame-link transit
from .packet import *
from .frame import FRAME_END, FRAME_ESC, FRAME_END_T, FRAME_ESC_T

class Parser:
    STATE_BUFFERING = 0
    STATE_ESCAPED = 1

    # requires a callback function that will be called when a packet is parsed
    # from incremental data
    def __init__(self, packet_callback):
        self.packet_callback = packet_callback
        self.clear()

    def clear(self):
        self.state = Parser.STATE_BUFFERING
        self.controlFlags = 0x0000
        self.frameBuffer = []
        
    def acceptPacket(self):
        packet = Packet(self.frameBuffer)
        err = None
        if (self.packet_callback):
            err = self.packet_callback(packet) # deliver to application code
        self.clear()
        return err

    # consumes data incrementally and emits calls to packet_callback when full
    # packets are parsed from the incoming stream
    def readBytes(self, buffer):
        for msg in buffer:
            if msg == FRAME_END:
                self.acceptPacket()
            elif self.state == Parser.STATE_ESCAPED:
                if msg == FRAME_END_T:
                    self.frameBuffer.append(FRAME_END)
                elif msg == FRAME_ESC_T:
                    self.frameBuffer.append(FRAME_ESC)                    
                self.state = Parser.STATE_BUFFERING
            elif (msg == FRAME_ESC):
                self.state = Parser.STATE_ESCAPED
            else:
                self.frameBuffer.append(msg)
